File: machinelearning/cpp/concept/finds.cpp
/**************************************************
filename       :finds.cpp
author         :maksyuki
created time   :2016年09月21日 星期三 08时16分16秒
last modified  :2016年09月21日 星期三 20时23分51秒
file location  :/home/maksyuki/Documents/github/learning/machinelearning/cpp
***************************************************/

#include <iostream>
#include <string>
using namespace std;

string data[16][16] = {
{"Example", "Sky",   "AirTemp", "Humidity", "Wind",   "Water", "Forecast", "EnjoySport"},
{"1",       "Sunny", "Warm",    "Normal",   "Strong", "Warm",  "Same",     "Yes"},
{"2",       "Sunny", "Warm",    "High",     "Strong", "Warm",  "Same",     "Yes"},
{"3",       "Rainy", "Cold",    "High",     "Strong", "Warm",  "Change",   "No"},
{"4",       "Sunny", "Warm",    "High",     "Strong", "Cool",  "Change",   "Yes"}
};

string ans[16];

void Init()
{
    for(int i = 0; i < 16; i++)
        ans[i] = "NULL";
}

void FindS()
{
    for(int i = 1; i < 5; i++)
    {
        if(data[i][7] == "Yes")
        {
            for(int j = 1; j < 7; j++)
            {
                if(ans[j] == "NULL") ans[j] = data[i][j];
                else if(ans[j] != data[i][j]) ans[j] = "?";
            }
        }
    }
}

int main()
{
    Init();
    FindS();
    for(int i = 1; i < 7; i++)
    {
        cout << data[0][i];
        if(i == 1)      cout << ":      ";
        else if(i == 2) cout << ":  ";
        else if(i == 4) cout << ":     ";
        else if(i == 5) cout << ":    ";
        else            cout << ": ";
        cout << "\"" << ans[i] << "\"" << endl;
    }

    return 0;
}

File: machinelearning/python/concept/finds.py
import re
from functools import reduce

input_string = """
Example Sky AirTemp Humidity Wind Water Forecast  EnjoySport
1       Sunny  Warm  Normal  Strong  Warm  Same   Yes
2       Sunny  Warm  High    Strong  Warm  Same   Yes
3       Rainy  Cold  High    Strong  Warm  Change No
4       Sunny  Warm  High    Strong  Cool  Change Yes"""
lines = list(map(lambda x: list(filter(lambda y: y != '',
                             re.split(' +', x)))[1:],  # drop first item "Example"
            filter(lambda x: x != '', re.split('\n', input_string))))
names, data = lines[0], lines[1:]
data_lines = list(map(lambda x: dict(zip(names, x)), data))
values = dict(zip(names,
                  reduce(lambda x, y:
                         map(lambda z:
                             z[0] | set([z[1]]) if type(z[0]) is set else set(z),
                             zip(x, y)),
                         lines[1:])))
names, target = names[:-1], names[-1]  # the last item is target
p = 'Yes'
n = 'No'

rule = reduce(lambda r, s:
              r if s[target] == n else
              dict(zip(names,
                       map(lambda name:
                           s[name] if r[name] is None else '?' if r[name] != s[name] else r[name],
                           names))),
              data_lines, dict.fromkeys(names, None))
print(rule)
